### PR TITLE
Add mypy, type hints to common.py

### PR DESCRIPTION
**Pull request by JevinJ**
TODO: 

- [x] Fix the cause of CI/tests failing, I think it's a circular dependency or something like that.
- [x] Fix style errors.
- [x] Choose a type checker pytype, mypy, etc. @akaszynski @banesullivan?
- [x] Resolve marked TODO's(in code) and verify correct types were chosen.
- [x] ~~I started this PR using typing for my types, but these are depreciated after 3.9. These should be replaced with collections versions.~~ Collections types aren't usable previous to 3.9

### CODE REVIEW

**Comment by akaszynski** [comment]
Dropbox is a big believer of `mypy`.  Let's go with that.

However, I think `mypy` integration will take a while:
```
...
pyvista/core/composite.py:143: error: Incompatible return value type (got "List[Any]", expected "Tuple[Any, ...]")
pyvista/core/composite.py:183: error: Signature of "get_data_range" incompatible with supertype "DataObject"
pyvista/core/composite.py:205: error: Bracketed expression "[...]" is not valid as a type
pyvista/core/composite.py:205: note: Did you mean "List[...]"?
pyvista/core/composite.py:250: error: Bracketed expression "[...]" is not valid as a type
pyvista/core/composite.py:250: note: Did you mean "List[...]"?
pyvista/core/composite.py:277: error: Incompatible return value type (got "List[Optional[str]]", expected "List[str]")
pyvista/core/composite.py:322: error: Bracketed expression "[...]" is not valid as a type
pyvista/core/composite.py:322: note: Did you mean "List[...]"?
Found 205 errors in 20 files (checked 45 source files)
```
@JevinJ, I recommend we do this by steps and not convert the entire module at once.  This is such a big PR that we won't be able to get it all in one go without merge conflicts either.  Let's do it in steps and incorporate `mypy` along the way so we know we're doing it correctly.

We can leave our current changes in (provided CI passes), and then run `mypy` on an isolated set of files.  For example:
```
mypy pyvista/core/composite.py
```

File: pyvista/plotting/helpers.py [hunk 1]
```diff
@@ -1,19 +1,38 @@
 """This module contains some convenience helper functions."""
+from typing import Optional, List, Union
 
+import itkwidgets
```
Comment (akaszynski) [inline pyvista/plotting/helpers.py:4]: We can't import this here as `itkwidgets` is an optional feature.

Comment (JevinJ) [reply pyvista/plotting/helpers.py:4]: I am aware, but it must be imported for type hints. Don't really have a good solution yet, could just leave this one out.

Comment (akaszynski) [reply pyvista/plotting/helpers.py:4]: `itkwidgets` doesn't have typing hints anyway, so let's not worry about this.

**Comment by akaszynski** [comment]
While I don't want to abandon this PR (again), I really think we need to focus on converting in small PRs.  Here's the rationale:
- Being a non-trivial PR, there's a ton to overwrite and we're already running into merge conflicts.
- We need to incorporate `mypy` into it at start, otherwise there will be a second huge rewrite when we then add in its support.
- Unit testing and CI must incorporate `mypy`, so when we contribute and merge, we are aware that it's passing.  We can add it as a separate CI step since it's quick and we want to know immediately if something has failed (like the code check).

Perhaps leave this branch up and then create a new PR that uses `mypy` for a single file.  Once that works, we can expand from there for 0.29.0

**Comment by JevinJ** [comment]
I just dropped all files except common.py and rebased, seems I accidentally overwrote the mypy.ini blame with my name at some point.

**Comment by JevinJ** [comment]
I've gotten mypy down to 7 errors. Not sure about the rest involving numpy/pyvista_ndarray.

**Comment by akaszynski** [comment]
> I've gotten mypy down to 7 errors. Not sure about the rest involving numpy/pyvista_ndarray.

I'll take a look at that this evening.

**Comment by JevinJ** [comment]
There is one more thing before this gets merged, PEP 8 specifies that there should be whitespace around '=' for optional parameters that use type hints. Should this be adhered to?

**Comment by akaszynski** [comment]
> Should this be adhered to? Always seemed contradictory and ugly to me.

Could you show me in the docs where you found this?  I'm having trouble finding it.

**Comment by JevinJ** [comment]
> 
> 
> > Should this be adhered to? Always seemed contradictory and ugly to me.
> 
> Could you show me in the docs where you found this? I'm having trouble finding it.

https://www.python.org/dev/peps/pep-0484/#arbitrary-argument-lists-and-default-argument-values
 
Shows this in examples, or:

https://docs.python.org/3/library/typing.html#typing.Optional

PyCharm complains about the no whitespace syntax.

**Comment by akaszynski** [comment]
Ran `mypy pyvista` and got:
```
Found 59 errors in 17 files (checked 44 source files)
```

Going to have to look at this tomorrow.

**Comment by akaszynski** [comment]
> Should this be adhered to? Always seemed contradictory and ugly to me.

I agree with PEP484.  Keyword args shouldn't have spaces, but when there is a space for type hints, it indicates that you're not assigning the value to the data type:
```py
def quux(__x: int, __y__: int = 0) -> None: ...
```

I'd be confused if `int=0`, as I'd think that something was assigned to `int` rather than to `__y__`.

While I'm all for breaking PEP when it doesn't make sense, most of the time they do and this is a case where I'm happy with it.

**Comment by akaszynski** [comment]
@JevinJ, this will be good to merge shortly, thanks for adding the type hints.  As our code base grows, I think enforcing type hints will be quite valuable as it will help us capture bugs where CI and inspection will fail.  It's a pain to implement, but it's a good move going forward.

One note when running `make mypy`, you will probably have to create a virtualenv to capture any type errors that show up in Azure CI but can't be duplicated locally.  This was really starting to drive me crazy when debugging, and even when ignoring the cache with `mypy pyvista/core/common.py --no-incremental`, I still couldn't duplicate the `mypy` errors unless I created a virtual environment, cloned the repo, and then ran `mypy`.

Moving forward for 0.29, we can add type hints one big module at a time to avoid merge hell as these PRs involve a ton of line changes.

**Review by akaszynski — APPROVED** [review]

**Pull request merged**